File: modules/rust_workloads/benches/bench_pointer_chase.rs
//! Pointer-chase bench: serially dependent random loads (memory-latency
//! bound) against a contiguous sum of the same buffer (bandwidth bound).

use std::hint::black_box;

use criterion::{Criterion, criterion_group, criterion_main};
use pprof::criterion::{Output, PProfProfiler};
use rust_workloads::{pointer_chase, workload_n};

const DEFAULT_N: usize = 1 << 22;

fn bench_pointer_chase(c: &mut Criterion) {
    let n = workload_n(DEFAULT_N);
    let perm = pointer_chase::build_cycle(n, 42);
    let mut group = c.benchmark_group("pointer_chase");
    group.bench_function("chase", |bench| {
        bench.iter(|| pointer_chase::chase_sum(black_box(&perm)));
    });
    group.bench_function("array_sum", |bench| {
        bench.iter(|| pointer_chase::array_sum(black_box(&perm)));
    });
    group.finish();
}

criterion_group! {
    name = benches;
    // 10 samples: one chase walk over the 32 MB cycle takes hundreds of ms, so
    // the default 100 samples cannot fit criterion's 5s measurement window.
    config = Criterion::default()
        .sample_size(10)
        .with_profiler(PProfProfiler::new(100, Output::Protobuf));
    targets = bench_pointer_chase
}
criterion_main!(benches);

File: modules/java_workloads/benches/BenchPointerChase.java
// Pointer-chase bench: serially dependent random loads (memory-latency
// bound) against a contiguous sum of the same buffer (bandwidth bound).

package monorepo.workloads;

import java.util.concurrent.TimeUnit;

import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.BenchmarkMode;
import org.openjdk.jmh.annotations.Fork;
import org.openjdk.jmh.annotations.Mode;
import org.openjdk.jmh.annotations.OutputTimeUnit;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.Setup;
import org.openjdk.jmh.annotations.State;

@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.MILLISECONDS)
@State(Scope.Benchmark)
@Fork(1)
public class BenchPointerChase {
    private static final int DEFAULT_N = 1 << 22;

    private int[] perm;

    @Setup
    public void setup() {
        perm = PointerChase.buildCycle(WorkloadN.workloadN(DEFAULT_N), 42);
    }

    @Benchmark
    public long chase() {
        return PointerChase.chaseSum(perm);
    }

    @Benchmark
    public long arraySum() {
        return PointerChase.arraySum(perm);
    }
}
